Process Listening On Port: store

erthalion opened the pull request:
Implement datastore for Process Listening On Port objects.

## Checklist
- [x] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
~- [ ] Evaluated and added CHANGELOG entry if required~
~- [ ] Determined and documented upgrade steps~
~- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

There are no breaking changes, no update steps (the PR is about handling a type
of data) and no user facing changes yet (they come in the follow up PRs).

## Testing Performed

Local testing, unit-tests, verifying previous test CI pipelines (e.g. #4182).
As nothing is using the datastore yet, no E2E testing is present yet.

erthalion committed:
Support various types of pruning for PLOP

There are two ways how pruning should work for PLOP:

* We need to clean up PLOP objects, when the corresponding Process
Indicator was pruned.

* We need to clean up PLOP objects without the process reference.

erthalion committed:
Process Listening On Port Storage

Implement datastore for PLOP objects, including both generated store,
datastore, needed metrics.

Co-authored-by: Jouko Virtanen <jvirtane@redhat.com>

erthalion committed:
Update definitions

Modify necessary proto definitions for PLOP and generate corresponding
files.

dashrews78 commented on central/processindicator/datastore/singleton.go:
```diff
@@ -33,20 +34,24 @@ var (
 	} else {
 		storage = rocksdb.New(globaldb.GetRocksDB())
+		// PLOP storage is only supported for PostgreSQL
+		// plopStorage = rocksdb.New(globaldb.GetRocksDB())
```
remove this line (or both)

JoukoVirtanen replied:
Done

dashrews78 commented on central/processlisteningonport/datastore/datastore.go:
```diff
@@ -0,0 +1,36 @@
+// DataStore interface for ProcessListeningOnPort object interaction with the database
+//
+//go:generate mockgen-wrapper
+type DataStore interface {
```
nit:  We should probably be consistent at least for the exported interface routines.
`AddProcessListeningOnPort`
`GetProcessListeningOnPort`
`RemovePLOP`

IMO we should either shorten them all or spell them all out.

JoukoVirtanen replied:
RemovePLOP was renamed to RemoveProcessListeningOnPort.

dashrews78 commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -0,0 +1,398 @@
+			indicatorID = indicator.GetId()
+			log.Debugf("Got indicator %s: %+v", indicatorID, indicator)
+		} else {
+			// XXX: Create a metric for this
```
nit:  should reference the Jira for adding such metric

JoukoVirtanen replied:
Changed to "TODO ROX-14377:"

dashrews78 commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -0,0 +1,398 @@
+			processInfo = val.Process
+		}
+
+		plopKey := fmt.Sprintf("%d %d %s",
```
nit:  This likely doesn't matter at all, but if we were to ever look at these keys visually it might be better to have a delimiter other than a space.  a `-` or `_` would be more obvious that it is a multiple items mashed together into 1 than a space does.

JoukoVirtanen replied:
We now have various functions to get the various keys and all keys are obtained by calling those functions. All of them use "_" as the delimiter.

dashrews78 commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -0,0 +1,398 @@
+		existingPLOP, prevExists := existingPLOPMap[plopKey]
+
+		// There are three options:
+		// * We found an existing PLOP object with different close timestamp.
```
what if the new close timestamp is before the existing close timestamp?  I don't think we guarantee these things arrive in order.

JoukoVirtanen replied:
We now sort the existing plops by CloseTimestamp. The following function has been added

```
func sortByCloseTimestamp(values []*storage.ProcessListeningOnPortFromSensor) {
	sort.Slice(values, func(i, j int) bool {
		return values[i].CloseTimestamp.Compare(values[j].CloseTimestamp) == -1
	})
}
```

We also have a test for plops being out of order in central/processlisteningonport/datastore/singleton.go. That test is TestPLOPAddCloseBatchOutOfOrder.

dashrews78 commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -0,0 +1,398 @@
+			newPLOP := &storage.ProcessListeningOnPortStorage{
+				// XXX, ResignatingFacepalm: Use regular GENERATE ALWAYS AS
+				// IDENTITY, which would require changes in store generator
+				Id:                 uuid.NewV4().String(),
```
we've probably talked around the edges of this before, but what do you want the `Id` to be?  I understand not wanting to fix the generator now and I get not wanting to make it a custom store now.  But we need to consider what a future migration of the primary key would be if we were to change its type.  It is possible to likely that migration may be just as cumbersome to deal with as taking on the task of making the PK what you want it to be now.  Something to consider.

erthalion replied:
My proposal was described in ROX-13793, to summarize the idea is to let the database to generate the id value. I know the migrations could be cumbersome in the future, but this one ideally should be easy, as it will be not a data migration (there still will be uuid), but only schema migration (with an extra logic how to generate uuid). Does it answer your concerns?

dashrews78 commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -0,0 +1,398 @@
+//
+// XXX: This function queries all PLOP, no matter if they are matching port +
+// protocol we've got or not. This means potentially dangerous corner cases
+// when one process listenst to a huge number of ports. To address it we could
```
nit:  typo `listenst`

JoukoVirtanen replied:
Good catch.

dashrews78 commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -0,0 +1,398 @@
+	// If no corresponding processes found, we can't verify if the PLOP
+	// object is opening/closing an existing one. Collect existingPLOPMap
+	// only if there are some matching indicators.
+	existingPLOP, err := ds.storage.GetByQuery(ctx, search.NewQueryBuilder().
```
recommend `existingPLOPs`

JoukoVirtanen replied:
Done

dashrews78 commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -0,0 +1,398 @@
+	}
+
+	for _, val := range existingPLOP {
+		key := fmt.Sprintf("%d %d %s",
```
same comment as earlier on the spaces.  Additionally we should probably create a helper function to create the key.

JoukoVirtanen replied:
See above response to similar comment.

dashrews78 commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -0,0 +1,398 @@
+		)
+
+		// A bit of paranoia is always good
+		if old, ok := existingPLOPMap[key]; ok {
```
Are they returned in order, so we know we want to keep the most recent one in the list?

JoukoVirtanen replied:
See above response to similar comment.

dashrews78 replied:
to clarify:  In the case of this function is it that we only ever have the most recent in the database as was calculated through the combination of normalize and Add functions?  Meaning this particular if check should not occur?

JoukoVirtanen replied:
Sorry for the confusion my previous response in this thread caused. I deleted it. We should not have entries with with identical keys here where the key consists of the protocol, port, and indicatorID. That is because of the normalize and Add functions. We should never see the log statement "A PLOP %s is already present, overwrite with %s"

dashrews78 commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -0,0 +1,398 @@
+	}
+
+	for _, val := range indicators {
+		key := fmt.Sprintf("%s %s %s %s %s",
```
same key comment on spaces

dashrews78 commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -0,0 +1,398 @@
+// same combination of port, protocol, process. Find and squash them into a
+// single event.
+//
+// Open/close state will be calculated from the totall number of open/close
```
typo `totall`

JoukoVirtanen replied:
Done

dashrews78 commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -0,0 +1,398 @@
+	normalizedResult := []*storage.ProcessListeningOnPortFromSensor{}
+
+	for _, val := range plops {
+		key := fmt.Sprintf("%d %d %s",
```
same key comment on spaces

JoukoVirtanen replied:
See above response to similar comment.

dashrews78 replied:
looks like this one is still built with spaces.

JoukoVirtanen replied:
Oops. Fixed now.

dashrews78 commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -0,0 +1,398 @@
+
+		// Take the last open PLOP if there are more open in total, otherwise
+		// the last closed PLOP
+		if nOpen > nClosed {
```
I might be missing something.  How does this handle the case of PLOPs that are out of order?  This seems to return the last one you processed which would assume that `plops` was passed in ordered.  May not matter so much for open, but could pose issues with closed and the closedTimestamp.

JoukoVirtanen replied:
See above response to similar comment.

dashrews78 commented on central/processlisteningonport/store/postgres/index.go:
```diff
@@ -0,0 +1,70 @@
+	"github.com/stackrox/rox/pkg/search/postgres/mapping"
+)
+
+func init() {
```
#4350 moves this to `pkg/postgres/schema`.  So watch for when that one is merged and adjust accordingly.

erthalion replied:
Already done, it was preventing the pipeline runs.

dashrews78 commented on central/processlisteningonport/store/postgres/store.go:
```diff
@@ -0,0 +1,748 @@
+// Code generated by pg-bindings generator.
```
nit:  `// Code originally generated by pg-bindings generator.`

dashrews78 replied:
My preference would be to do it now and get it over with so it doesn't linger.  It also clears it up that the store is in fact no longer generated.  But again that is my opinion.

erthalion replied:
I have to disagree here. Having this done will introduce unnecessary divergence between generated version and generated bits of the custom version, making it much more painful to reconcile them in the future. I mean, how silly would that be when we will get conflicts just because of `podId` will be different from `podID`. And the fast that the store is not longer generated is stated clearly everywhere else.

If you have a strong opinion on this topic, I would be fine following your suggestion, but will recruit you in the future for reconciliation work :)

dashrews78 replied:
One thing to note, such divergence is already unavoidable as #4210 resolved a lot of the style issues for these generated stores.  that specifically targeted things like changing podId to be podID.  of course an option could be to save off the custom bits, regenerate with #4210 and then add the custom stuff back.

dashrews78 commented on central/processlisteningonport/store/postgres/store.go:
```diff
@@ -0,0 +1,748 @@
+	targetResource = resources.DeploymentExtension
+)
+
+type Store interface {
```
nit:  #4210 reordered the routines into a logical order.  Up to you if you want to mimic that.

erthalion replied:
We have generated those bits again to apply the new ordering coming from this change.

dashrews78 commented on central/pruning/pruning_test.go:
```diff
@@ -567,7 +567,7 @@ func (s *PruningTestSuite) TestImagePruning() {
 			alerts, config, images, deployments, pods, indexQ := s.generateImageDataStructures(ctx)
```
Need to add a test for PLOP pruning.

JoukoVirtanen replied:
We have added three tests for PLOP pruning.

Test that open PLOP is not pruned.
Test that closed PLOP which is not expired is not pruned.
Test that closed PLOP which is expired is pruned.

dashrews78 commented on pkg/postgres/schema/process_indicators.go:
```diff
@@ -1,4 +1,7 @@
-// Code generated by pg-bindings generator. DO NOT EDIT.
+// Code generated by pg-bindings generator.
```
nit:  `// Code originally generated by pg-bindings generator.`

JoukoVirtanen replied:
Making this change causes style-check to run on this file, and that reports style errors for automatically generated code.

JoukoVirtanen replied:
Done

dashrews78 commented on pkg/postgres/schema/process_indicators.go:
```diff
@@ -45,16 +48,20 @@ const (
 )
 
 // ProcessIndicators holds the Gorm model for Postgres table `process_indicators`.
+// XXX: unique_indicator index is specified manually, as it has to cover
```
If you have the Jira ticket to do this later, reference it.

JoukoVirtanen committed:
Add tests for batch processing

Concentrace on corner cases: events completed inside the batch, ourder
of events, etc.

erthalion committed:
Batch normalization improvements

Fix various corner cases for batch normalization:

* Avoid situations when adding nil plop object (do not allocate upfront,
append instead)
* Separate normalized events into normal and those that are completed
inside the batch
* Sort event inside the batch by CloseTimestamp for out-of-order events

JoukoVirtanen committed:
Added unit tests for plop pruning

It covers PLOP objects pruning only, not involving pruning that happens
as part of Process Indicator pruning.

erthalion commented:
/retest-required

JoukoVirtanen committed:
Address comments the PR review

Collection of various small improvements to address the PR review.

rhybrillou commented on central/pruning/pruning_test.go:
```diff
@@ -1307,6 +1308,100 @@ func (s *PruningTestSuite) TestRemoveOrphanedProcesses() {
+			expectedDeletions: []string{},
+		},
+		{
+			name: "Plop is closed but not expried so it is not removed",
```
nit: typo in `expired` ?

JoukoVirtanen replied:
Fixed

rhybrillou commented on pkg/postgres/schema/process_indicators.go:
```diff
@@ -47,16 +50,20 @@ const (
 	DeploymentId       string `gorm:"column:deploymentid;type:uuid;index:processindicators_deploymentid,type:hash"`
-	ContainerName      string `gorm:"column:containername;type:varchar"`
-	PodId              string `gorm:"column:podid;type:varchar"`
+	ContainerName      string `gorm:"column:containername;type:varchar;index:unique_indicator"`
```
Could the index name be prefixed similarly to the `processindicators_poduid` one ?

erthalion replied:
Sure, makes sense.

JoukoVirtanen commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -0,0 +1,398 @@
+			val.GetProtocol(),
+			val.GetPort(),
+			indicatorID,
+		)
```
We now have various functions to get the various keys and all keys are obtained by calling those functions.

erthalion commented on pkg/postgres/schema/process_indicators.go:
```diff
@@ -47,16 +50,20 @@ const (
-	ContainerName      string `gorm:"column:containername;type:varchar"`
-	PodId              string `gorm:"column:podid;type:varchar"`
+	ContainerName      string `gorm:"column:containername;type:varchar;index:unique_indicator"`
+	PodId              string `gorm:"column:podid;type:varchar;index:unique_indicator"`
```
Yep, we've already created ROX-13790 for that.

erthalion committed:
Regenerate store

Generator has being changed, impacting e.g. the order of methods. To be
as close as the generated version, regenerate once more and apply custom
changes on top.

erthalion commented on central/processindicator/store/postgres/gen.go:
```diff
@@ -1,3 +0,0 @@
-package postgres
-
-//go:generate pg-table-bindings-wrapper --type=storage.ProcessIndicator --search-category PROCESS_INDICATORS --references storage.Deployment
```
I'm a bit cautious about doing this, after we've learned that removing the header will make those files eligible for style checking, which they not always pass. In case of process indicator schema it looks fine for now, but maybe it will not be like this in the future. What do you think, how important is it to update the header, having in mind the clear statement about generation status later on and those practical considerations I've mentioned above?

dashrews78 replied:
I think my preference would be to update this stuff on the process indicators now as painful as it may be.  My reasoning is that we have no idea what else may need to change in process indicators outside of your PLOP work over the next few sprints and we would want it to be as obvious as possible to other contributors as to what is occurring.  That is just one person's opinion of course.

erthalion commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -0,0 +1,398 @@
+) (
+	processesListeningOnPorts []*storage.ProcessListeningOnPort, err error,
+) {
+	if ok, err := plopSAC.ReadAllowed(ctx); err != nil {
```
To be honest I'm not familiar enough with the scope access mechanism, could you show an example how to modify `ReadAllowed` call to check ClusterID and Namespace as well?

erthalion replied:
What happened here is that there is a possibility to do SAC verification more fine grained, and allow users who have only access to certain namespaces to see PLOPs. As in example [[1]]  we would need to fetch cluster id and namespace for that. The problems would be to do this efficiently via deployment argument that we have in this API, as we only have the deployment name, which is not an indexed column in the db. All in all we agree that this could be done as a follow up, and needs to be investigated first in ROX-14417.

[1]: https://github.com/stackrox/stackrox/blob/master/central/deployment/datastore/datastore_impl.go#L129-L140

dashrews78 commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -0,0 +1,503 @@
+		if indicator, ok := indicatorsMap[key]; ok {
+			indicatorID = indicator.GetId()
+			log.Debugf("Got indicator %s: %+v", indicatorID, indicator)
+		} else {
```
do we ever clean up and/or reconcile the case where there is no indicatorID?  Theoretically that could result in a fair number of collisions with the map and orphaned data in the database.

erthalion replied:
They're getting cleaned up during pruning. There are ideas about reconciliation, but first we wanted to get understanding how frequent this happening in practice.

erthalion committed:
Rename unique index for Process Indicator

To be consistent with other names use processindicators prefix.

dashrews78 commented on central/processlisteningonport/datastore/datastore_impl_test.go:
```diff
@@ -0,0 +1,1149 @@
+}
+
+func (suite *PLOPDataStoreTestSuite) SetupTest() {
+	if !env.PostgresDatastoreEnabled.BooleanSetting() {
```
if we use a `SetupSuite` we should be able to do this check one time and then call a `suite.T().Skip()` or something like that and skip the entire suite so we don't have to put this check in each test.  Something like in `central/node/datastore/dackbox/datastore/datastore_impl_test.go` only the reverse case.

JoukoVirtanen replied:
Done

dashrews78 commented on central/processlisteningonport/datastore/datastore_impl_test.go:
```diff
@@ -0,0 +1,1149 @@
+		suite.hasWriteCtx, fixtureconsts.Deployment1)
+	suite.NoError(err)
+
+	// It's being closed and excluded from the API response
```
Comment seems wrong since it has been re-opened.

JoukoVirtanen replied:
It has been corrected

dashrews78 commented on pkg/postgres/schema/process_listening_on_ports.go:
```diff
@@ -0,0 +1,57 @@
+// Code generated by pg-bindings generator. DO NOT EDIT.
```
Is this one still being generated?  the line in `gen.go` is commented out so I suspect it is not.

erthalion replied:
`gen.go` for PLOP is still in place, only one for ProcessIndicator was removed. PLOP doesn't generate the store itself due to `--schema-only`, but the db schema is still automated and in sync.

dashrews78 commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -0,0 +1,503 @@
+		return sac.ErrResourceAccessDenied
+	}
+
+	portProcesses, completedInBatch := normalizePLOPs(portProcesses)
```
Do we really want to reuse the name of the incoming parameter?  I would think we would want to name it something else to differentiate since we pass that into `normalizePLOPs` and get 2 things back.

JoukoVirtanen replied:
No. portProcesses has been changed to normalizedPLOPs

JoukoVirtanen commented on central/pruning/pruning.go:
```diff
@@ -417,6 +426,41 @@ func (g *garbageCollectorImpl) removeOrphanedProcessBaselines(deployments set.Fr
+	// This walk seems unnecessary, but ClosedTimestamp is a part of serialized
+	// column, not visible to Postgres. So if we would like to keep using
+	// orphanWindow, unfortunately it has to be this way, at least for now.
+	walkFn := func() error {
```
Done

JoukoVirtanen commented on central/processindicator/datastore/datastore.go:
```diff
@@ -43,9 +44,10 @@ type DataStore interface {
 
 // New returns a new instance of DataStore using the input store, indexer, and searcher.
-func New(store store.Store, indexer index.Indexer, searcher search.Searcher, prunerFactory pruner.Factory) (DataStore, error) {
+func New(store store.Store, plopStore plopStore.Store, indexer index.Indexer, searcher search.Searcher, prunerFactory pruner.Factory) (DataStore, error) {
```
Done

JoukoVirtanen commented on central/processindicator/datastore/datastore_impl.go:
```diff
@@ -137,12 +142,28 @@ func (ds *datastoreImpl) removeMatchingIndicators(ctx context.Context, results [
 }
 
 func (ds *datastoreImpl) removeIndicators(ctx context.Context, ids []string) error {
+	log.Infof("Process Pruning")
```
Removed

JoukoVirtanen commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -0,0 +1,398 @@
+}
+
+func (ds *datastoreImpl) removePLOP(ctx context.Context, ids []string) error {
+	log.Infof("Deleting PLOP records")
```
Done

JoukoVirtanen commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -0,0 +1,398 @@
+	if len(ids) == 0 {
+		return nil
+	}
+	if err := ds.storage.DeleteMany(ctx, ids); err != nil {
```
Done

JoukoVirtanen commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -0,0 +1,398 @@
+func (ds *datastoreImpl) fetchExistingPLOPs(
+	ctx context.Context,
+	indicatorIds []string,
+	portProcesses ...*storage.ProcessListeningOnPortFromSensor,
```
Done

JoukoVirtanen commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -0,0 +1,398 @@
+			val.GetSignal().GetName(),
+			val.GetSignal().GetArgs(),
+			val.GetSignal().GetExecFilePath(),
+		)
```
Done

JoukoVirtanen commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -0,0 +1,485 @@
+func addNewPLOP(plopObjects []*storage.ProcessListeningOnPortStorage,
+	indicatorID string,
+	processInfo *storage.ProcessIndicatorUniqueKey,
+	value *storage.ProcessListeningOnPortFromSensor) []*storage.ProcessListeningOnPortStorage {
```
Done

JoukoVirtanen committed:
Remove "Code generated by..." headers for custom code

To minimize chances for confusion and make it crystal clear which parts
are not automatically generated any more, adjust generation headers for
corresponding files. This means we also have to apply style fixes as
well.

JoukoVirtanen committed:
Skipping tests in SetupSuite if postgres is not used

erthalion committed:
Sanity check for PLOP DataStore singletone

dashrews78 commented on central/processindicator/store/postgres/store.go:
```diff
@@ -1,4 +1,4 @@
-// Code generated by pg-bindings generator. DO NOT EDIT.
+// Code originally generated by pg-bindings generator.
```
nit:  Can we add a TODO with the reference to the ticket(s) we plan to use to get this back to being generated?

JoukoVirtanen commented on central/processlisteningonport/store/postgres/index.go:
```diff
@@ -0,0 +1,65 @@
+// Code generated by pg-bindings generator. DO NOT EDIT.
```
These comments have been changed to "Code originally generated by pg-bindings generator"

rhybrillou commented on central/processlisteningonport/store/postgres/store.go:
```diff
@@ -0,0 +1,763 @@
+		}
+
+		var msg storage.ProcessListeningOnPortStorage
+		if err := proto.Unmarshal(serialized, &msg); err != nil {
```
If I remember correctly, this pattern was replaced in `pkg/search/postgres/common.go` to leverage the `Unmarshal` function in the generated code. The change was driven by performance considerations (https://github.com/stackrox/stackrox/pull/3451).
https://github.com/stackrox/stackrox/blob/master/pkg/search/postgres/common.go#L883-L893
I think the pattern here would be 
```
    msg := &storage.ProcessListeningOnPortStorage{}
    if err := msg.Unmarshal(serialized); err != nil {
        return nil, err
    }
```
This change is not critical in my eyes, applies to the ProcessIndicator decoding as well, and can be the target of a follow-up PR.

dashrews78 commented on central/processlisteningonport/datastore/datastore_impl_test.go:
```diff
@@ -0,0 +1,1100 @@
+}
+
+func (suite *PLOPDataStoreTestSuite) SetupTest() {
+	suite.hasNoneCtx = sac.WithGlobalAccessScopeChecker(context.Background(), sac.DenyAllAccessScopeChecker())
```
nit:  If you don't need this stuff to be reset for each test, you can move it to `SetupSuite` if you want.

erthalion committed:
Replace custom PLOP store with a wrapper

Create a wrapper around generated PLOP store, called full store, to
incorporate both auto generated and custom bits together. The advantage
is that both live in separate files, allowing to continue auto generate
the main store part.

rhybrillou commented on central/processlisteningonport/store/postgres/store.go:
```diff
@@ -0,0 +1,763 @@
+		"FROM process_listening_on_ports plop " +
+		"JOIN process_indicators proc " +
+		"ON plop.processindicatorid = proc.id " +
+		"WHERE proc.deploymentid = $1 AND plop.closed = false"
```
AFAIK, In the current state of things, the pattern applied in our app is to duplicate the data from the serialized object in the table columns that can be used for searching.
As a consequence, I think the query here could simply select `plop.id, plop.serialized, proc.serialized as procSerialzed`. This would slightly reduce the amount of data sent over the wire, and decoded by the process.

erthalion replied:
It's probably fine to use this approach for this particular PR, if everything else works this way. But in long term perspective such duplication is a huge problem (more data to store, bulletproof mechanism is required to make sure columns are really the same as in serialized field) and has to be addressed.

rhybrillou commented on central/processlisteningonport/store/postgres/store.go:
```diff
@@ -0,0 +1,763 @@
+
+		plop := &storage.ProcessListeningOnPort{
+			Endpoint: &storage.ProcessListeningOnPort_Endpoint{
+				Port:     msg.Port,
```
General consideration applicable for all fields used to build the plop: the getters from the generated protobuf code should be nil-safe.
I'd prefer here `msg.GetPort()` to `msg.Port`, or later `procMsg.GetSignal().GetId()` to `procMsg.Signal.Id`.

JoukoVirtanen replied:
Done

rhybrillou commented on central/processlisteningonport/store/postgres/store.go:
```diff
@@ -0,0 +1,763 @@
+				Protocol: msg.Protocol,
+			},
+			DeploymentId:  procMsg.DeploymentId,
+			PodId:         podID,
```
If the suggested simplified query statement were used, `podID` would not have to be decoded, and `procMsg.GetPodId()` could fill the same purpose.
Variation of this could be applied for `procMsg.GetContainerName()`, `procMsg.GetSignal().GetContainerId()` and the other select fields that could be deduced from the serialized payload.

JoukoVirtanen replied:
Done

rhybrillou reviewed (commented):
Minor comments on the store part.

JoukoVirtanen committed:
Using serialized fields

It turns out that serialized column duplicates all the rest of the
columns, in this big scheme of things use it for now.

JoukoVirtanen committed:
Moved sac from SetupTest to SetupSuite

erthalion commented:
/retest-required

JoukoVirtanen commented:
/retest

JoukoVirtanen commented:
/retest

rhybrillou commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -0,0 +1,503 @@
+	}
+
+	// Verify what to do about pairs of open/close events that close the
+	// lifecycle withih the batch. There are only few options:
```
typo: within

JoukoVirtanen replied:
Done

rhybrillou commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -0,0 +1,503 @@
+	}
+
+	// Now save actual PLOP objects
+	err = ds.storage.UpsertMany(ctx, plopObjects)
```
nit: `return ds.storage.UpsertMany(ctx, plopObjects)`

JoukoVirtanen replied:
Done

rhybrillou commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -0,0 +1,503 @@
+	var existingPLOPMap = map[string]*storage.ProcessListeningOnPortStorage{}
+
+	if len(indicatorIds) == 0 {
+		return nil, nil
```
I'd rather return the empty map than nil here.

JoukoVirtanen replied:
Done

erthalion committed:
Introduce necessary changes to ProcessIndicator

Some queries for PLOP has to be supported by ProcessIndicator table
schema. Extend it with necessary index, temporarily making it custom,
i.e. not generated via storage generator.

JoukoVirtanen committed:
Tests improvements

* Verify Process column more closely (for those PLOP events that end up
in completed bucket after batchnig).
* Add few utility functions for fetching content of the tables, and use them for checking expected values.
* Make sure the db is clear before actually running tests.
* Remove unused bits.
* Correct few commentaries.
* Remove unneed test with identical Process Indicators (they were not
really identical, so the test didn't make much sense).

erthalion committed:
Fix ineffassign

rhybrillou commented on central/processlisteningonport/datastore/datastore_impl.go:
```diff
@@ -0,0 +1,498 @@
+
+func sortByCloseTimestamp(values []*storage.ProcessListeningOnPortFromSensor) {
+	sort.Slice(values, func(i, j int) bool {
+		return values[i].CloseTimestamp.Compare(values[j].CloseTimestamp) == -1
```
(ideally for both compared entries)

erthalion committed:
Collection of various improvements

* Add new unit test for our of order PLOP events for batching, this time
with more "open" events. Consequently, it makes the final result "open".
* Clean up full_store implementation from unused bits.
* Do not warn on NoRows in custom store functions.
* Tiny style changes.

erthalion commented on central/processlisteningonport/datastore/datastore_impl_test.go:
```diff
@@ -0,0 +1,1099 @@
+	suite.NoError(err)
+	suite.Len(plopsFromDB, 1)
+	suite.Equal(plopsFromDB[0].CloseTimestamp, closedPlopObject3.CloseTimestamp)
+}
```
Yeah, why not. Added a new test for that as well.

JoukoVirtanen commented:
/test gke-postgres-nongroovy-e2e-tests

JoukoVirtanen commented:
/test gke-kernel-qa-e2e-tests
/test openshift-newest-qa-e2e-tests
/test openshift-oldest-qa-e2e-tests
